### PR TITLE
New encoder

### PR DESCRIPTION
**Pull request by bitanath**
Added new encoder and deleted the earlier branch to prevent Mac artifacts like MacOSX and .DS_Store

### CODE REVIEW

File: encoder.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
```
Comment (rosasalberto) [inline encoder.py:4]: delete all these comments

File: encoder.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+import matplotlib.pylab as plt
+import json
+
+import os
+import subprocess
+import re
+import sys
+import imageio
+import argparse
+
```
Comment (rosasalberto) [inline encoder.py:43]: check PEP guide for imports.
```
Imports should be grouped in the following order:

Standard library imports.
Related third party imports.
Local application/library specific imports.
You should put a blank line between each group of imports.
```

File: encoder.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
+#!/usr/bin/env python
```
Comment (rosasalberto) [inline encoder.py:1]: Use OOP to modularize the different code functionalities

File: encoder.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+detector = dlib.get_frontal_face_detector()
+shape_predictor = dlib.shape_predictor("weights/shape_predictor_68_face_landmarks.dat")
+vgg16=VGG16(include_top=False,input_shape=(224,224,3),weights='imagenet')
+vggmod=Model(vgg16.input, vgg16.layers[6].output)
+truncation_psi = 0.5
+w_average = np.load('weights/{}_dlatent_avg.npy'.format(weights))
+generator = StyleGan2Generator(weights="ffhq", impl=impl, gpu=args.use_gpu)
+
+
```
Comment (rosasalberto) [inline encoder.py:75]: all things related to landmark detection should be in another class. e.g. LandmarkDetector

File: encoder.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+# In[4]:
+
+
+vgg = VGG(vggmod)
+vgg.trainable = False
+
+
+# In[5]:
```
Comment (rosasalberto) [inline encoder.py:95]: The definition of the initializer (in your case the imagenet model) should be in another class. E.g Initializer, where you import the model defined in other file.

File: encoder.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+def generateFromWP(wp):
+    wp = tf.convert_to_tensor(np.repeat(wp.reshape((1,512)),18,0).reshape((1,18,512)))
+    dlatents = w_average + (wp - w_average) * truncation_psi
+    out = generator.synthesis_network(dlatents)
+    int8 = convert_images_to_uint8(out, nchw_to_nhwc=True, uint8_cast=True)
+    img = Image.fromarray(int8.numpy()[0])
+    return img
+
+def displaySideBySide(im1,im2):
```
Comment (rosasalberto) [inline encoder.py:113]: this function should be in utils

File: encoder.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+# In[5]:
+
+
+def wpize(n):
+    rnd = np.random.RandomState(n)
+    z = rnd.randn(1, 512).reshape(1,512)
+    w = generator.mapping_network(z.astype('float32'))
+    wp = np.array(w[0][1]).reshape(1,512)
+    return wp
+
```
Comment (rosasalberto) [inline encoder.py:104]: this 2 functions should be methods in the Encoder class, with more descriptive names

File: encoder.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+def generateFromWP(wp):
```
Comment (rosasalberto) [inline encoder.py:105]: make sure to look at the PEP style for naming convention, in this case, methods or functions should use lower case with underscore.

File: encoder.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+    plt.imshow(np.rot90(im1,4))
+    f.add_subplot(1,2, 2)
+    plt.imshow(np.rot90(im2,4))
+    plt.show(block=True)
+
+
+# In[6]:
+
+
+def align(img, output_size=1024, transform_size=4096, enable_padding=True, x_scale=1, y_scale=1, em_scale=0.1, alpha=False):
```
Comment (rosasalberto) [inline encoder.py:125]: this should be in another class. E.g. ImageAligner

File: encoder.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            blur = qsize * 0.02
+            img += (scipy.ndimage.gaussian_filter(img, [blur, blur, 0]) - img) * np.clip(mask * 3.0 + 1.0, 0.0, 1.0)
+            img += (np.median(img, axis=(0,1)) - img) * np.clip(mask, 0.0, 1.0)
+            img = np.uint8(np.clip(np.rint(img), 0, 255))
+            if alpha:
+                mask = 1-np.clip(3.0 * mask, 0.0, 1.0)
+                mask = np.uint8(np.clip(np.rint(mask*255), 0, 255))
+                img = np.concatenate((img, mask), axis=2)
+                img = Image.fromarray(img, 'RGBA')
+            else:
```
Comment (rosasalberto) [inline encoder.py:202]: Split this function into multiple other functions depending on their functionality. 
Make sure to check CS SOLID principles

File: encoder.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        if output_size < transform_size:
+            img = img.resize((output_size, output_size), Image.ANTIALIAS)
+
+        return img
+
+
+# In[10]:
+
+
+raw = Image.open(filename)
```
Comment (rosasalberto) [inline encoder.py:217]: open Images in other classes

File: encoder.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+enco = np.random.randn(1, 512).reshape(1,512)
+label = vgg(preprocess_input(np.array(image.resize((224,224))).reshape(1,224,224,3)))
+vggweights = vggmod.get_weights()[-1]
+lr_schedule = tf.keras.optimizers.schedules.ExponentialDecay(initial_learning_rate=0.100,decay_steps=100,decay_rate=0.9)
+optimizer = Adam(learning_rate=lr_schedule,amsgrad=False) 
+
+
+# In[ ]:
+
+
```
Comment (rosasalberto) [inline encoder.py:228]: The optimization process should be in another class for example. LatentOptimizer

**Review by rosasalberto — COMMENTED** [review]
Hi @bitanath,
Thanks for the PR. 
Check the comments I made. 
It will be interesting if you can train the initializer doing the inverse mapping, and adding the pre-trained weights for that.
Contact me via email (alberto.rosas@upc.edu) if you have any doubt.

**Comment by bitanath** [comment]
Got you, thanks @rosasalberto , I'll work on these over the weekend and resubmit

**Comment by johndpope** [comment]
so Nvidia labs will be using pytorch from now on - there's never going to be an official tensorflow 2 - stylegan2 library.
https://github.com/NVlabs/stylegan2-ada/issues/32

when I look at their code and derivative works - there's a lot of people who have implemented things on top of the dnn folder.
Is it possible to just auto upgrade it automagically - or is your approach / complete rewrite better?
I'm stuck on latest 3090 card and can't get code to run without docker. 

I guess - what I'm wanting to say is - can you help port the nvidia labs code to tensorflow2? 
I want to use existing trained pickels - this code does a lot to convert tensorflow pickels - to pytorch.
maybe helpful - https://github.com/GreenLimeSia/GenEdi

UPDATE
I created / cherry picked a v2 compatibility branch here 
https://github.com/johndpope/stylegan2-ada 

https://github.com/johndpope/stylegan2-ada/tree/digressions

**Pull request closed without merging**